woocommerce/woocommerce-gateway-stripe
# Add SetupIntents so the customer can confirm an SCA-protected payment even on a free order

This PR replaces https://github.com/woocommerce/woocommerce-gateway-stripe/pull/961 and https://github.com/woocommerce/woocommerce-gateway-stripe/pull/997 . Really, it's on top of both those PRs, but because the extensive behaviour change that has been done, I figured this would look better in a single PR, instead of having #961 implement something that this PR would undo.

This PR introduces SetupIntents so the user has a chance to pass a SCA challenge even if the charge is not going to be charged right away. That's important for Subscriptions that have a free trial period, and pre-orders.

Passing the SCA challenge in this way makes it less probable that passing another challenge will be required down the road. We are assuming that most credit cards will behave like this Stripe test card (which is the one I recommend to test this BTW):

NUMBER | AUTHENTICATION | DESCRIPTION
-- | -- | --
4000002500003155 | Required on setup or first transaction | This test card requires authentication for one-time payments. However, if you set up this card using the Setup Intents API and use the saved card for subsequent payments, no further authentication is needed.

The big change for this PR is that **IF THE SCA CHALLENGE IS NOT PASSED, THE ORDER IS MARKED AS FAILED**. So, to sum up, this "optional" SetupIntent SCA challenge is treated as mandatory.

To test:
- Create a subscription with free trial.
- Sign up for that subscription, go to the checkout.
- Use `4242424242424242` credit card, see that the order is created normally and the subscription too.
- Sign up for a subscription again, this time using the `4000002500003155` card. After clicking the `Sign Up Now` button at checkout, you'll get an SCA challenge modal indicating that this is a "non-payment authentication test page". At that moment, the order is in a "Pending payment" status, and so the subscription is not yet active.
- If you pass the challenge, you'll be redirected to the "Thank you" page. The subscription will be active at this point.
- If you fail the challenge, you'll stay in the checkout with an error message, and you'll be able to try again. At this point, the order status is "Failed" and so the subscription is not active.

PS: I haven't made sure that Pre-Orders work, because I found some weird stuff in the compatibility class. I'll take a deeper look tomorrow.

Fixes #936
